File: update_checker.py
"""Background update checker for SOSterm.

Checks https://sos-tech.ca/updates/fterm/latest.json for new versions.
Runs in a QThread to avoid blocking the UI. Respects a 24-hour cooldown
between automatic checks.
"""

import json
import time
import urllib.request
import urllib.error
from PyQt5.QtCore import QThread, pyqtSignal

UPDATE_URL = "https://sos-tech.ca/updates/fterm/latest.json"
CHECK_COOLDOWN = 86400  # 24 hours in seconds


def compare_versions(current, remote):
    """Return True if remote is newer than current (semver comparison)."""
    try:
        cur = [int(x) for x in current.split(".")]
        rem = [int(x) for x in remote.split(".")]
        return rem > cur
    except (ValueError, AttributeError):
        return False


class UpdateCheckThread(QThread):
    """Background thread that checks for updates."""

    update_available = pyqtSignal(str, str, str)  # version, download_url, changelog
    check_finished = pyqtSignal(bool)  # had_update

    def __init__(self, current_version, parent=None):
        super().__init__(parent)
        self._current_version = current_version

    def run(self):
        try:
            req = urllib.request.Request(
                UPDATE_URL,
                headers={"User-Agent": f"SOSterm/{self._current_version}"},
            )
            with urllib.request.urlopen(req, timeout=10) as resp:
                data = json.loads(resp.read().decode("utf-8"))

            remote_version = data.get("version", "")
            download_url = data.get("download_url", "")
            changelog = data.get("changelog", "")

            if compare_versions(self._current_version, remote_version):
                self.update_available.emit(remote_version, download_url, changelog)
                self.check_finished.emit(True)
            else:
                self.check_finished.emit(False)
        except (urllib.error.URLError, json.JSONDecodeError, OSError, KeyError):
            self.check_finished.emit(False)


class UpdateChecker:
    """Manages update checks with cooldown logic."""

    def __init__(self, current_version, settings):
        self._current_version = current_version
        self._settings = settings
        self._thread = None

    def should_auto_check(self):
        """Return True if enough time has passed since last check."""
        last = self._settings.get("last_update_check", 0)
        return (time.time() - last) >= CHECK_COOLDOWN

    def check(self, on_update=None, on_finished=None, record_time=True):
        """Start a background update check.

        on_update: callback(version, download_url, changelog)
        on_finished: callback(had_update)
        """
        if self._thread and self._thread.isRunning():
            return

        self._thread = UpdateCheckThread(self._current_version)
        if on_update:
            self._thread.update_available.connect(on_update)
        if on_finished:
            self._thread.check_finished.connect(on_finished)
        if record_time:
            self._settings.set("last_update_check", time.time())
        self._thread.start()

    def auto_check(self, on_update=None, on_finished=None):
        """Check only if cooldown has elapsed."""
        if self.should_auto_check():
            self.check(on_update=on_update, on_finished=on_finished)

File: src/updatechecker.cpp
#include "updatechecker.h"
#include "settings.h"

#include <QDateTime>
#include <QJsonDocument>
#include <QJsonObject>
#include <QNetworkAccessManager>
#include <QNetworkReply>
#include <QNetworkRequest>

static const char *UPDATE_URL = "https://sos-tech.ca/updates/fterm/latest.json";
static const int COOLDOWN_SECS = 86400; // 24 hours

UpdateChecker::UpdateChecker(Settings *settings, const QString &currentVersion,
                             QObject *parent)
    : QObject(parent)
    , m_nam(new QNetworkAccessManager(this))
    , m_settings(settings)
    , m_currentVersion(currentVersion)
{
}

bool UpdateChecker::shouldAutoCheck() const
{
    qint64 last = m_settings->get("last_update_check", 0).toLongLong();
    qint64 now = QDateTime::currentSecsSinceEpoch();
    return (now - last) >= COOLDOWN_SECS;
}

void UpdateChecker::check(bool recordTime)
{
    m_recordTime = recordTime;

    QNetworkRequest req{QUrl(QString::fromLatin1(UPDATE_URL))};
    req.setHeader(QNetworkRequest::UserAgentHeader,
                  QStringLiteral("SOSterm/%1").arg(m_currentVersion));

    auto *reply = m_nam->get(req);
    connect(reply, &QNetworkReply::finished, this, &UpdateChecker::onReplyFinished);
}

void UpdateChecker::autoCheck()
{
    if (shouldAutoCheck())
        check(true);
}

void UpdateChecker::onReplyFinished()
{
    auto *reply = qobject_cast<QNetworkReply *>(sender());
    if (!reply)
        return;
    reply->deleteLater();

    if (m_recordTime)
        m_settings->set("last_update_check",
                        (qlonglong)QDateTime::currentSecsSinceEpoch());

    if (reply->error() != QNetworkReply::NoError) {
        emit checkFinished(false);
        return;
    }

    QJsonDocument doc = QJsonDocument::fromJson(reply->readAll());
    if (!doc.isObject()) {
        emit checkFinished(false);
        return;
    }

    QJsonObject obj = doc.object();
    QString remoteVersion = obj["version"].toString();
    QString downloadUrl = obj["download_url"].toString();
    QString changelog = obj["changelog"].toString();

    if (isNewer(m_currentVersion, remoteVersion)) {
        emit updateAvailable(remoteVersion, downloadUrl, changelog);
        emit checkFinished(true);
    } else {
        emit checkFinished(false);
    }
}

bool UpdateChecker::isNewer(const QString &current, const QString &remote)
{
    auto parse = [](const QString &v) -> QVector<int> {
        QVector<int> parts;
        for (const auto &s : v.split('.'))
            parts.append(s.toInt());
        return parts;
    };

    QVector<int> cur = parse(current);
    QVector<int> rem = parse(remote);

    int len = qMax(cur.size(), rem.size());
    for (int i = 0; i < len; ++i) {
        int c = i < cur.size() ? cur[i] : 0;
        int r = i < rem.size() ? rem[i] : 0;
        if (r > c) return true;
        if (r < c) return false;
    }
    return false;
}
